src/lib/operators/join_mpsm.cpp

```diff
@@ -0,0 +1,520 @@
+        auto new_pos_list = _dereference_pos_list(input_table, column_id, pos_list);
+
+        if (input_table->chunk_count() > 0) {
+          const auto base_column = input_table->get_chunk(ChunkID{0})->get_column(column_id);
+          const auto ref_column = std::dynamic_pointer_cast<const ReferenceColumn>(base_column);
+
+          auto new_ref_column = std::make_shared<ReferenceColumn>(ref_column->referenced_table(),
+                                                                  ref_column->referenced_column_id(), new_pos_list);
+          output_columns.push_back(new_ref_column);
+        } else {
+          // If there are no Chunks in the input_table, we can't deduce the Table that input_table is referencING to
+          // pos_list will contain only NULL_ROW_IDs anyway, so it doesn't matter which Table the ReferenceColumn that
+          // we output is referencing. HACK, but works fine: we create a dummy table and let the ReferenceColumn ref
+          // it.
+          const auto dummy_table = Table::create_dummy_table(input_table->column_definitions());
+          output_columns.push_back(std::make_shared<ReferenceColumn>(dummy_table, column_id, pos_list));
+        }
+      } else {
+        auto new_ref_column = std::make_shared<ReferenceColumn>(input_table, column_id, pos_list);
+        output_columns.push_back(new_ref_column);
+      }
+    }
+  }
+
+  /**
+  * Turns a pos list that is pointing to reference column entries into a pos list pointing to the original table.
+  * This is done because there should not be any reference columns referencing reference columns.
+  **/
+  std::shared_ptr<PosList> _dereference_pos_list(std::shared_ptr<const Table> input_table, ColumnID column_id,
+                                                 std::shared_ptr<const PosList> pos_list) {
+    // Get all the input pos lists so that we only have to pointer cast the columns once
+    auto input_pos_lists = std::vector<std::shared_ptr<const PosList>>();
+    for (ChunkID chunk_id{0}; chunk_id < input_table->chunk_count(); ++chunk_id) {
+      auto b_column = input_table->get_chunk(chunk_id)->get_column(column_id);
+      auto r_column = std::dynamic_pointer_cast<const ReferenceColumn>(b_column);
+      input_pos_lists.push_back(r_column->pos_list());
+    }
+
+    // Get the row ids that are referenced
+    auto new_pos_list = std::make_shared<PosList>();
+    for (const auto& row : *pos_list) {
+      if (row.chunk_offset == INVALID_CHUNK_OFFSET) {
+        new_pos_list->push_back(RowID{ChunkID{0}, INVALID_CHUNK_OFFSET});
+      } else {
+        new_pos_list->push_back((*input_pos_lists[row.chunk_id])[row.chunk_offset]);
+      }
+    }
+
+    return new_pos_list;
+  }
+
+ public:
+  /**
+  * Executes the MPSMJoin operator.
+  **/
+  std::shared_ptr<const Table> _on_execute() {
+    bool include_null_left = (_mode == JoinMode::Left || _mode == JoinMode::Outer);
```

Comment:
The JoinHash swaps the input tables for RightOuterJoins so it only has to support left outer joins. Maybe this could reduce the complexity of you code too, just speculating

Reply:
The problem with switching is that it makes estimating the cost of joins more complicated as the optimizer needs to guess whether the join switches or not (this is not necessarily a trivial decision).
Hence, we decided to just adhere to the interface and keep the left column as the left column (and accept the added complexity in the join). 

As a consequence, the optimizer needs to select the correct order. But as the optimizer needs to account for that anyways (e.g., use hash join when on table is significantly smaller), we think this is a fair assumption.